pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -14,227 +14,202 @@
+    # Recover the operators in the decomposition; note that because of the
+    # initial SWAP, we exchange the order of A and B
+    A_ops = zyz_decomposition(A, wires[1])
+    B_ops = zyz_decomposition(B, wires[0])
+    C_ops = zyz_decomposition(C, wires[0])
+    D_ops = zyz_decomposition(D, wires[1])
 
+    return C_ops + D_ops + [qml.CNOT(wires=[wires[0], wires[1]])] + A_ops + B_ops
 
-def _extract_su2su2_prefactors(U, V):
-    r"""U, V are SU(4) matrices for which there exists A, B, C, D such that
-    (A \otimes B) V (C \otimes D) = U. The problem is to find A, B, C, D in SU(2)
-    in an analytic and fully differentiable manner.
 
-    This decomposition is possible when U and V are in the same double coset of
-    SU(4), meaning there exists G, H in SO(4) s.t. G (Edag V E) H = (Edag U
-    E). This is guaranteed here by how V was constructed using the
-    _select_rotation_angles method. Then, we can use the fact that E SO(4) Edag
-    gives us something in SU(2) x SU(2) to give A, B, C, D.
+def _decomposition_2_cnots(U, wires):
+    r"""If 2 CNOTs are required, we can write the circuit as
+     -U- = -A--X--RZ(d)--X--C-
+     -U- = -B--C--RX(p)--C--D-
+    We need to find the angles for the Z and X rotations such that the inner
+    part has the same spectrum as U, and then we can recover A, B, C, D.
     """
+    # Compute the rotation angles
+    u = math.dot(Edag, math.dot(U, E))
+    gammaU = math.dot(u, math.T(u))
+    evs, _ = math.linalg.eig(gammaU)
 
-    # A lot of the work here happens in the magic basis. Essentially, we
-    # don't look explicitly at some U = G V H, but rather at
-    #     E^\dagger U E = G E^\dagger V E H
-    # so that we can recover
-    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D).
-    # There is some math in the paper explaining how when we define U in this way,
-    # we can simultaneously diagonalize functions of U and V to ensure they are
-    # in the same coset and recover the decomposition.
-
-    u = math.dot(math.cast_like(Edag, V), math.dot(U, math.cast_like(E, V)))
-    v = math.dot(math.cast_like(Edag, V), math.dot(V, math.cast_like(E, V)))
-
-    uuT = math.dot(u, math.T(u))
-    vvT = math.dot(v, math.T(v))
-
-    # First, we find a matrix p (hopefully) in SO(4) s.t. p^T u u^T p is diagonal.
-    # Since uuT is complex and symmetric, both its real / imag parts share a set
-    # of real-valued eigenvectors.
-    if math.get_interface(u) == "tensorflow":
-        ev_p, p = math.linalg.eig(math.real(uuT))
+    # These choices are based on Proposition III.3 of
+    # https://arxiv.org/abs/quant-ph/0308045
+    # There is, however, a special case where the circuit has the form
+    # -U- = -A--C--X--C-
+    # -U- = -B--X--C--D-
+    #
+    # or some variant of this, where the two CNOTs are adjacent.
+    #
+    # What happens here is that the set of evs is -1, -1, 1, 1 and we can write
+    # -U- = -A--X-SZ-X--C-
+    # -U- = -B--C-SX-C--D-
```

Comment:
I'm a bit confused here; how come the top circuit shows adjacent CNOTs, but the bottom one no-longer has adjacent CNOTs?

Reply:
Ah, this is a very cute circuit transformation! So what's happened is that for the first CNOT, you can flip it's direction by surrounding it with Hadamards:
```
 -A--C--X--C- = -A--H--X--H--X--C-
 -B--X--C--D- = -B--H--C--H--C--D-
```

Then, you decompose the H on the top into `[SX, S, SX]`, and the one on the bottom into `[S, SX, S]`:
```
-A--H--X--H--X--C- = -A--H--X--SX-S--SX--X--C-
-B--H--C--H--C--D- = -B--H--C--S--SX--S--C--D
```
Then you can pull the SX through the targets of the CNOTs on the top because they commute, and squish all the single-qubit gates together. And similarly, the S on the bottom can be pulled through the controls, and combine single-qubit gates:
```
-A--H--SX--X--S---X--SX-C- = -A'--X--S--X--C'--
-B--H--S---C--SX--C--S--D- = -B'--C--SX-C--D'--
```

Reply:
Neat! Thanks for the explanation